specs/GetSecurityDescriptorForPackageFamilyNames/GetSecurityDescriptorForPackageFamilyNames.md

```diff
@@ -0,0 +1,49 @@
+# Easing and streamlining named object sharing between packaged and Win32
+
+This feature is a flat C API that takes a list of Package Family Names (PFNs) and access masks, and returns a security descriptor
+
+# Background
+
+Named objects (mutexes, events, waitable timers) must be ACLed during creation in order to be shared with packaged applications.
+
+Currently, [our guidance](https://docs.microsoft.com/en-us/windows/uwp/communication/sharing-named-objects) for this scenario is to write a large amount of complex, cumbersome, and largely boilerplate code. ([See here for a taste](https://docs.microsoft.com/en-us/windows/win32/api/securityappcontainer/nf-securityappcontainer-getappcontainernamedobjectpath#examples)).
+
+If we are to meet developers where they are, we need to reduce the friction of Win32 and UWP interop, and make it as easy as possible to share named objects.
+
+# Description
+
+Add a flat C API that takes an array of PFNs and access masks, and produces the needed `SECURITY_ATTRIBUTES` object.
+
+# Examples
+
+```c++
+#include <wil/resource.h>
+
+wil::unique_event CreateShareableEvent(PCWSTR name)
+{
+    wil::unique_hlocal_ptr pSD;
+    PCWSTR packageNames[1] = { L"Contoso.Test.App_12345678" };
+    DWORD accessMasks[1] = { ACCESS_MODIFY_STATE };
+
+    THROW_IF_FAILED(GetSecurityDescriptorForPackageFamilyNames(
+        1, packageNames, accessMasks, &pSD);
+
+    wil::unique_event result;
+    result.create(wil::EventOptions::None, name, pSD.get());
+    return result;
+}
+```
+
+# API Details
+
+```c
+STDAPI GetSecurityDescriptorForPackageFamilyNames(
+    DWORD                 cCountOfPackageFamilyNames,
+    _In_Reads_(cCountOfPackageFamilyNames)
+      const PCWSTR*         pListOfPackageFamilyNames,
+    _In_Reads_(cCountOfPackageFamilyNames)
+      const DWORD*          accessMask,
+    _Outptr_ SECURITY_DESCRIPTOR** ppSD
```

Comment:
The type expected by CreateNamedPipe and friends is LPSECURITY_ATTRIBUTES. I think this API should return that type (which has the security descriptor inside of it). The main goal here is to make the simple majority case trivial and not require the developer to learn a complicated security model API or have to create a bunch of nested structs.

Reply:
A SECURITY_ATTRIBUTES describes _how_ a security descriptor should be used.  It seems reasonable to add helper/wrapper methods like "CreateSharedEvent" or "SetSharedKeySecurity" or whatever, while retaining this primtive.

Reply:
The main goal is to make it easy to generate the DACL and pass it to the right Create* function. Is creating the SECURITY_ATTRIBUTES struct trivial once you have the security descriptor? What does that code look like? Presumably you just set the security descriptor field, set inherit to false(?), and what do you set for nLength?

Reply:
I've updated the example code to show this. It's pretty trivial
```c++
    SECURITY_ATTRIBUTES sa;
    sa.nLength = sizeof(SECURITY_ATTRIBUTES);
    sa.lpSecurityDescriptor = pSD.get();
    sa.bInheritHandle = FALSE;
```

Reply:
Ok. I'm ok with that as long as we document this along with the API itself and how to use it.

Reply:
Would this be equivalent
```
SECURITY_ATTRIBUTES sa{ sizeof(sa), pSD.get() };
```
?

Reply:
Howard, already wrote this, prefer sizeof(sa) over the named type (less error prone).